dlib/CMakeLists.txt

```diff
@@ -810,21 +797,22 @@ if (NOT TARGET dlib)
                cuda/cuda_data_ptr.cpp
                cuda/gpu_data.cpp
                )
-            list (APPEND dlib_needed_private_libraries ${CUDA_CUBLAS_LIBRARIES})
+            list (APPEND dlib_needed_private_libraries CUDA::toolkit)
+            list (APPEND dlib_needed_private_libraries CUDA::cublas)
             list (APPEND dlib_needed_private_libraries ${cudnn})
-            list (APPEND dlib_needed_private_libraries ${CUDA_curand_LIBRARY})
-            list (APPEND dlib_needed_private_libraries ${CUDA_cusolver_LIBRARY})
-            list (APPEND dlib_needed_private_libraries ${CUDA_CUDART_LIBRARY})
+            list (APPEND dlib_needed_private_libraries CUDA::curand)
+            list (APPEND dlib_needed_private_libraries CUDA::cusolver)
+            list (APPEND dlib_needed_private_libraries CUDA::cudart)
```

Comment:
This won't work when generating pkgconfig files.
If you look at my old PR https://github.com/davisking/dlib/pull/2912, which included using FindCUDAToolkit.cmake, I was unable to properly interoperate with pkgconfig due to cmake's use of "blah::blah" when linking libraries. This was my final hurdle but sadly couldn't overcome it. It included a whole bunch of other tidy-ups too... Sad

Reply:
Oh no, this sounds bad. I wasn't reading all of the PR, because it is way too long. But I assume you reference this comment [1] of yours.

You suggest using the "old" cmake way of using XXX_LIBRARIES and XXX_INCLUDE_DIRS.
Given the result variables from the new finder [2], the major problem seems to be the XXX_LIBRARIES variable, which is not set. I was also browsing the cmake source code of the finder and there isn't even a private one.

Did you try out to get all the link libraries via a target property? [3] Maybe creating the XXX_LIBRARIES manually by extracting all the libraries from the LINK_LIBRARIES [4] property is fine.


[1] https://github.com/davisking/dlib/pull/2912#issuecomment-1975149618
[2] https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#result-variables
[3] get_target_property: https://cmake.org/cmake/help/latest/command/get_target_property.html
[4] LINK_LIBRARIES: https://cmake.org/cmake/help/latest/prop_tgt/LINK_LIBRARIES.html

Reply:
I tried extracting all the link libraries but it isn't trivial. There are some old cmake libraries that attempt to do this but they're not maintained. Personally my preference would be to drop support for pkgconfig then this problem goes away. There is a ticket open https://gitlab.kitware.com/cmake/cmake/-/issues/22621 that discusses adding support for exporting pkgconfig files natively. But nothing's happened. Realistically, how many users are using dlib via pkgconfig? My guess is not very many.

Reply:
The issue with the link libraries is that you have to do it recursively for all dependencies. It can get a bit nasty.

Reply:
Thanks, this sounds at least like a solution if pkgconfig cannot be dropped.

Long-term wise, a solution is probably needed anyway, as CMake linking on targets will very likely stay.

Reply:
I think we should add to CI/CD building an example program linking to dlib via pkgconfig.